market/mk20/types.go

```diff
@@ -0,0 +1,244 @@
+type FormatCar struct {
+	Version uint64 `json:"version"`
+}
+
+// FormatAggregate represents the aggregated format for piece data, identified by its type.
+type FormatAggregate struct {
+
+	// Type specifies the type of aggregation for data pieces, represented by an AggregateType value.
+	Type AggregateType `json:"type"`
+
+	// Sub holds a slice of PieceDataFormat, representing various formats of piece data aggregated under this format.
+	// The order must be same as segment index to avoid incorrect indexing of sub pieces in an aggregate
+	Sub []PieceDataFormat `json:"sub"`
+}
+
+// FormatBytes defines the raw byte representation of data as a format.
+type FormatBytes struct{}
+
+// DataSourceOffline represents the data source for offline pieces, including metadata such as the raw size of the piece.
+type DataSourceOffline struct {
+	RawSize uint64 `json:"rawsize"`
+}
+
+func (dso *DataSourceOffline) Name() DataSourceName {
+	return DataSourceNameOffline
+}
+
+func (dso *DataSourceOffline) IsEnabled(dbDataSources []dbDataSource) (int, error) {
+	name := string(dso.Name())
+	for _, p := range dbDataSources {
+		if p.Name == name {
+			if p.Enabled {
+				return Ok, nil
+			}
+		}
+	}
+	return ErrUnsupportedDataSource, xerrors.Errorf("data source %s is not enabled", name)
+}
+
+// DataSourceAggregate represents an aggregated data source containing multiple individual DataSource pieces.
+type DataSourceAggregate struct {
+	Pieces []DataSource `json:"pieces"`
+}
+
+func (dsa *DataSourceAggregate) Name() DataSourceName {
+	return DataSourceNameAggregate
+}
+
+func (dsa *DataSourceAggregate) IsEnabled(dbDataSources []dbDataSource) (int, error) {
+	name := string(dsa.Name())
+	for _, p := range dbDataSources {
+		if p.Name == name {
+			if p.Enabled {
+				return Ok, nil
+			}
+		}
+	}
+	return ErrUnsupportedDataSource, xerrors.Errorf("data source %s is not enabled", name)
+}
+
+// DataSourceHTTP represents an HTTP-based data source for retrieving piece data, including its raw size and associated URLs.
+type DataSourceHTTP struct {
+
+	// RawSize specifies the raw size of the data in bytes.
+	RawSize uint64 `json:"rawsize"`
+
+	// URLs lists the HTTP endpoints where the piece data can be fetched.
+	URLs []HttpUrl `json:"urls"`
+}
+
+func (dsh *DataSourceHTTP) Name() DataSourceName {
+	return DataSourceNameHTTP
+}
+
+func (dsh *DataSourceHTTP) IsEnabled(dbDataSources []dbDataSource) (int, error) {
+	name := string(dsh.Name())
+	for _, p := range dbDataSources {
+		if p.Name == name {
+			if p.Enabled {
+				return Ok, nil
+			}
+		}
+	}
+	return ErrUnsupportedDataSource, xerrors.Errorf("data source %s is not enabled", name)
+}
+
+// HttpUrl represents an HTTP endpoint configuration for fetching piece data.
+type HttpUrl struct {
+
+	// URL specifies the HTTP endpoint where the piece data can be fetched.
+	URL string `json:"url"`
+
+	// HTTPHeaders represents the HTTP headers associated with the URL.
+	HTTPHeaders http.Header `json:"httpheaders"`
```

Comment:
The struct is already named http, no need for that in variable names

```suggestion
	Headers http.Header `json:"headers"`
```